server-common/src/main/java/org/apache/kafka/server/common/DeleteRecordsPartitionStatus.java

```diff
@@ -0,0 +1,159 @@
+            super(delayMs);
+            this.onAcksPending = onAcksPending;
+            this.deleteRecordsStatus = new ConcurrentHashMap<>(deleteRecordsStatus);
+            this.responseCallback = responseCallback;
+            // first update the acks pending variable according to the error code
+            deleteRecordsStatus.forEach((topicPartition, status) -> {
+                if (status.responseStatus().errorCode() == Errors.NONE.code()) {
+                    // Timeout error state will be cleared when required acks are received
+                    status.setAcksPending(true);
+                    status.responseStatus().setErrorCode(Errors.REQUEST_TIMED_OUT.code());
+                } else {
+                    status.setAcksPending(false);
+                }
+            });
+        }
+
+        /**
+         * The delayed delete records operation can be completed if every partition specified in the request satisfied one of the following:
+         *
+         * 1) There was an error while checking if all replicas have caught up to the deleteRecordsOffset: set an error in response
+         * 2) The low watermark of the partition has caught up to the deleteRecordsOffset. set the low watermark in response
+         *
+         */
+        @Override
+        public boolean tryComplete() {
+            //  check for each partition if it still has pending acks
+            deleteRecordsStatus.forEach((topicPartition, status) -> {
+                log.trace("Checking delete records satisfaction for {}, current status {}", topicPartition, status);
+                //  skip those partitions that have already been satisfied
+                if (status.acksPending()) {
+                    onAcksPending.accept(topicPartition, status);
+                }
+            });
+            //  check if every partition has satisfied at least one of case A or B
+            return deleteRecordsStatus.values().stream().noneMatch(DeleteRecordsPartitionStatus::acksPending) && forceComplete();
+        }
+
+        @Override
+        public void onExpiration() {
+            deleteRecordsStatus.forEach((topicPartition, status) -> {
+                if (status.acksPending()) {
+                    DelayedDeleteRecordsMetrics.recordExpiration(topicPartition);
+                }
+            });
+        }
+
+        /**
+         * Upon completion, return the current response status along with the error code per partition
+         */
+        @Override
+        public void onComplete() {
+            Map<TopicPartition, DeleteRecordsResponseData.DeleteRecordsPartitionResult> responseStatus =
+                    new HashMap<>();
+            deleteRecordsStatus.forEach((k, status) -> {
+                responseStatus.put(k, status.responseStatus());
+            });
+            responseCallback.accept(responseStatus);
+        }
+    }
+
+    private static class DelayedDeleteRecordsMetrics {
```

Comment:
we don't need this static class as all static variables can be moved to `DelayedDeleteRecords`